src/wp-includes/pluggable.php

```diff
@@ -91,18 +94,25 @@ function get_userdata( $user_id ) {
 	 *
 	 * @since 2.8.0
 	 * @since 4.4.0 Added 'ID' as an alias of 'id' for the `$field` parameter.
+	 * @since 5.7.0 Returns the global $current_user if it's the user being fetched.
 	 *
 	 * @param string     $field The field to retrieve the user with. id | ID | slug | email | login.
 	 * @param int|string $value A value for $field. A user ID, slug, email address, or login name.
 	 * @return WP_User|false WP_User object on success, false on failure.
 	 */
 	function get_user_by( $field, $value ) {
+		global $current_user;
+
 		$userdata = WP_User::get_data_by( $field, $value );
 
 		if ( ! $userdata ) {
 			return false;
 		}
 
+		if ( $current_user instanceof WP_User && $current_user->ID == $userdata->ID ) {
```

Comment:
Does this run too late given the `get_data_by()` call above? (Genuine question not some horrible passive aggressive change request ;)

```suggestion
		if ( $current_user instanceof WP_User && $current_user->ID === $userdata->ID ) {
```